### PR TITLE
Long net Integration

### PR DESCRIPTION
**Pull request by JacksonSearle**
We updated torchscale. It now includes LongNet (as coded up by torchscale people). The readme has newer setup instructions for the new version of torchscale. There are two new arguments in the .yaml that are LongNet specific. Model types can still be swapped out by changing the model_type string in the .yaml file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,6 +108,7 @@ To run the grid search, ensure your configuration file is correctly set up, then
 
 ```bash
 python3 ../../grid_search.py configs/user_configs/<YOUR_CONFIG_HERE>.yaml
+python3 ../../grid_search.py configs/user_configs/<YOUR_CONFIG_HERE>.yaml
```
Comment (KimballNJardine) [inline README.md]: Remove duplicate line - probably came from merge conflict resolution by accident.

File: pip_install.sh
Comment (KimballNJardine) [inline pip_install.sh:1]: Why is this needed as a sh file? Please explain.

File: src/models.py [hunk 2]
```diff
@@ -318,6 +319,153 @@ def save_pretrained(self, save_folder: str):
...
+            decoder_value_embed_dim=config.value_embed_dim,
+            decoder_attention_heads=config.heads,
+            decoder_ffn_embed_dim=config.ffn_dim,
+            decoder_layers=config.layers,
+            dropout=config.dropout,
+            activation_dropout=config.activation_dropout,
+            vocab_size=config.vocab_size,
+            segment_length=config.segment_length,
+            dilated_ratio=config.dilated_ratio,
+            flash_attention=True)
```
Comment (KimballNJardine) [inline src/models.py]: Probably add a comment alerting readers to the fact that this is a constant, and does not come from a config file.

File: src/models.py [hunk 3]
```diff
@@ -318,6 +319,153 @@ def save_pretrained(self, save_folder: str):
...
+            on_epoch=True,
+            sync_dist=True,
+            add_dataloader_idx=True)
+
+        return loss
+
+    def test_step(self, batch: Tensor, batch_idx: int):
+        """ Test step, called automatically by PyTorch Lightning. """
+        # Unpack batch
+        inputs = batch[:, :-1]
```
Comment (KimballNJardine) [inline src/models.py]: Transformers architecture simply uses `inputs, targets = batch`.

File: src/models.py [hunk 4]
```diff
@@ -318,6 +319,153 @@ def save_pretrained(self, save_folder: str):
...
+            A tensor of dimensions: (batch size, sequence length, vocabulary
+                size).
+        """
+        preds = self.model_hf(x)
+        return preds
+
+    def training_step(self, batch: Tensor, batch_idx: int):
+        """ Training step, called automatically by PyTorch Lightning. """
+        # Unpack batch
+        inputs = batch[:, :-1]
```
Comment (KimballNJardine) [inline src/models.py]: Transformers uses `inputs, targets = batch`.

File: src/models.py [hunk 5]
```diff
@@ -318,6 +319,153 @@ def save_pretrained(self, save_folder: str):
...
+
+        self.log(
+            name="train_loss",
+            value=loss,
+            prog_bar=True,
+            logger=True,
+            on_step=True,
+            on_epoch=True,
+            sync_dist=True,
+            add_dataloader_idx=True)
```
Comment (KimballNJardine) [inline src/models.py:392]: Perplexity also needs to be calculated and logged.

File: src/models.py [hunk 6]
```diff
@@ -318,6 +319,153 @@ def save_pretrained(self, save_folder: str):
...
+
+        return loss
+
+    def validation_step(self, batch: Tensor, batch_idx: int):
+        """ Validation step, called automatically by PyTorch Lightning. """
+        # Unpack batch
+        inputs = batch[:, :-1]
+        targets = batch[:, 1:]
```
Comment (KimballNJardine) [inline src/models.py]: Update to use `inputs, targets = batch`.

File: src/torchscale/architecture/decoder.py [hunk 7]
```diff
@@ -155,6 +155,7 @@ def forward(
             attn_mask=self_attn_mask,
             rel_pos=self_attn_rel_pos,
             is_first_step=is_first_step,
+            is_causal=True,
```
Comment (KimballNJardine) [inline src/torchscale/architecture/decoder.py:158]: Please leave comment that this is hard-coded, unless there is a good reason not to.

File: pip_install.sh
Comment (nprisbrey) [inline pip_install.sh:1]: @JacksonSearle This file can now be removed, correct?

Comment (JacksonSearle) [reply pip_install.sh:1]: oh yeah, I accidentally left it there, give me a minute, I'll push it rn

Comment (JacksonSearle) [reply pip_install.sh:1]: k its gone now

File: src/torchscale/component/dilated_attention.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            qi = self.gathering(q, dr, sl, is_causal=is_causal, offset=offset, is_kv=False, seq_parall=self.args.seq_parallel)
+
+            out, lse = self.attention_ops(qi, ki, vi, key_padding_mask=key_padding_mask, attn_mask=attn_mask, rel_pos=rel_pos, is_causal=is_causal)
+
+            outs.append(out)
+            lses.append(lse)
+
+        attn = self.scattering(outs, lses, tgt_len, bsz, offset=offset)
+
+        # Convert data back to float32, since inner_attn_ln only works with float32
```
Comment (KimballNJardine) [inline src/torchscale/component/dilated_attention.py:212]: Note: lines 212-214 are only changes in this file compared to Microsoft Torchscale.

File: src/torchscale/component/embedding.py
Comment (KimballNJardine) [inline src/torchscale/component/embedding.py:1]: No difference with Microsoft torchscale embedding.py.

**Comment by nprisbrey** [comment]
@JacksonSearle, when @KimballNJardine and I were trying to follow the instructions to the letter on the README to run the LongNet model, we ran into issues on the supercomputer when we tried to train the model. There was an issue with the flash attention v2 and CUDA that prevented things from running ([this is the job output file](https://github.com/DRAGNLabs/301r_retnet/files/14681570/slurm_error.txt)).
We then tried to get around possible PyTorch, CUDA, and `flash-attn` version mismatches by double checking the version of CUDA in the environment and the job node, but this doesn't seem to be an issue since both were running on CUDA version 12.4. We then "loosened" the `torch` version in the `requirements.txt` file but ran into different errors when trying to install `flash-attn` with the newest version of `torch` ([this is the error we received from the `pip install flash-attn --no-build-isolation` command](https://github.com/DRAGNLabs/301r_retnet/files/14681569/flash-attn_install_error.txt)).
We're not sure how to recreate the environment that you ran the model in to test this ourselves. Could you give more details about steps to take/package versions to download to recreate the environment you had success in?

**Comment by JacksonSearle** [comment]
@nprisbrey Here's some info:

(retnet) (base) bash-4.2$ pip list
Package                  Version
------------------------ ------------
absl-py                  2.0.0
accelerate               0.26.1
aiohttp                  3.8.4
aiosignal                1.3.1
async-timeout            4.0.2
attrs                    23.1.0
cachetools               5.3.2
certifi                  2023.11.17
chardet                  5.2.0
charset-normalizer       3.3.2
click                    8.1.7
colorama                 0.4.6
contourpy                1.2.0
cycler                   0.12.1
DataProperty             1.0.1
datasets                 2.16.0
dill                     0.3.6
einops                   0.7.0
evaluate                 0.4.1
fairscale                0.4.13
filelock                 3.13.1
flash-attn               2.5.6
fonttools                4.47.2
frozenlist               1.3.3
fsspec                   2023.6.0
google-auth              2.26.2
google-auth-oauthlib     1.2.0
grpcio                   1.60.0
huggingface-hub          0.20.2
idna                     3.6
Jinja2                   3.1.2
joblib                   1.3.2
jsonlines                4.0.0
kiwisolver               1.4.5
lightning-utilities      0.10.1
lm_eval                  0.4.0
lxml                     5.1.0
Markdown                 3.5.2
MarkupSafe               2.1.3
matplotlib               3.8.2
mbstrdecoder             1.1.3
mpmath                   1.3.0
multidict                6.0.4
multiprocess             0.70.14
networkx                 3.2.1
ninja                    1.11.1.1
nltk                     3.8.1
numexpr                  2.9.0
numpy                    1.26.2
nvidia-cublas-cu12       12.1.3.1
nvidia-cuda-cupti-cu12   12.1.105
nvidia-cuda-nvrtc-cu12   12.1.105
nvidia-cuda-runtime-cu12 12.1.105
nvidia-cudnn-cu12        8.9.2.26
nvidia-cufft-cu12        11.0.2.54
nvidia-curand-cu12       10.3.2.106
nvidia-cusolver-cu12     11.4.5.107
nvidia-cusparse-cu12     12.1.0.106
nvidia-nccl-cu12         2.18.1
nvidia-nvjitlink-cu12    12.3.101
nvidia-nvtx-cu12         12.1.105
oauthlib                 3.2.2
packaging                23.2
pandas                   2.0.2
pathvalidate             3.2.0
peft                     0.8.2
Pillow                   10.1.0
pip                      23.3.1
portalocker              2.8.2
protobuf                 4.23.4
psutil                   5.9.8
pyarrow                  12.0.1
pyarrow-hotfix           0.6
pyasn1                   0.5.1
pyasn1-modules           0.3.0
pybind11                 2.11.1
pyparsing                3.1.1
pytablewriter            1.2.0
python-dateutil          2.8.2
pytorch-lightning        2.1.3
pytz                     2023.3.post1
PyYAML                   6.0.1
regex                    2023.6.3
requests                 2.31.0
requests-oauthlib        1.3.1
responses                0.18.0
rouge-score              0.1.2
rsa                      4.9
sacrebleu                2.4.0
safetensors              0.3.1
scikit-learn             1.4.0
scipy                    1.12.0
setuptools               68.0.0
six                      1.16.0
sqlitedict               2.1.0
sympy                    1.12
tabledata                1.3.3
tabulate                 0.9.0
tcolorpy                 0.1.4
tensorboard              2.15.1
tensorboard-data-server  0.7.2
threadpoolctl            3.2.0
timm                     0.9.12
tokenizers               0.15.0
torch                    2.1.1
torchdata                0.7.1
torchinfo                1.8.0
torchmetrics             1.3.1
torchtext                0.16.1
torchvision              0.16.1
tqdm                     4.66.1
tqdm-multiprocess        0.0.11
transformers             4.36.2
triton                   2.1.0
typepy                   1.3.2
typing_extensions        4.6.3
tzdata                   2023.3
urllib3                  2.1.0
Werkzeug                 3.0.1
wheel                    0.41.2
xxhash                   3.2.0
yarl                     1.9.2
zstandard                0.22.0
(retnet) (base) bash-4.2$ cuda --version
bash: cuda: command not found
(retnet) (base) bash-4.2$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Tue_Feb_27_16:19:38_PST_2024
Cuda compilation tools, release 12.4, V12.4.99
Build cuda_12.4.r12.4/compiler.33961263_0
(retnet) (base) bash-4.2$ python --version
Python 3.11.5

**Comment by nprisbrey** [comment]
@JacksonSearle does the most recent version of the code still work?

**Comment by JacksonSearle** [comment]
Hey, sorry, I was working on getting my .yaml up to date with the new changes in main. I had a couple of wrong file paths, not all the required arguments, etc. I'll take a look at it later this afternoon again

**Comment by nprisbrey** [comment]
@JacksonSearle, any update?

**Review by KimballNJardine — APPROVED** [review]
I could run LongNet, as well as RetNet and Transformer successfully. Hence, it works without breaking changes.
Approved.
